### PR TITLE
Add train loop support for looped PP schedules

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #411
* __->__ #358

----

- refactor some per-model logic into helper functions

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -250,6 +259,7 @@ def loss_fn(pred, labels):
     # train loop
     model.train()
 
+    # TODO(whc) Make CheckpointManager accept lists of models, optimizers, schedulers but save a single logical ckpt from them
```
Comment (fegin) [inline train.py]: So you will need a list of tuple, `List[Tuple[nn.Module, torch.optim.Optimizer, Scheduler]]`. That is important because we need to know which model is optimized with which optimizer to do the parameter ID -> FQN conversion.

Comment (wconstab) [reply train.py]: makes sense.
i guess its still possible to accept 3 separate tuples/lists at the CheckpointManager api, and assert same length, then just match them like 0:0:0, 1:1:1, 2:2:2?

or i need to keep them in a tuple together to give to DCP for some reason

Comment (fegin) [reply train.py]: yes, you can use 3 lists/tuples as long as CheckpointManager pass the correct pairs to the APIs.

File: train.py [hunk 2]
```diff
@@ -239,6 +247,10 @@ def loss_fn(pred, labels):
         f"({gpu_mem_stats.max_reserved_pct:.2f}%)"
     )
 
+    # TODO(whc) Make a wrapper class for Optimizers, Schedulers, expose 'step' API for convenience
```
Comment (awgu) [inline train.py]: I wonder what is feeling around using an `nn.ModuleList` instead of a normal Python `list` so that we can still treat it as a single root module.

Comment (wconstab) [reply train.py]: feels morally wrong to me. the modules are not really connected sequentially, and calling the .forward of the modulelist would be semantically invalid.

Comment (wconstab) [reply train.py]: otoh, it is a nice hack in many ways. it would let us wrap one optimizer, easily pass to DCP, (well, scratch that: it'd change fqn's on us wouldn't it?)

Comment (awgu) [reply train.py]: `nn.ModuleList` does not implement `forward` (unlike `nn.Sequential`), so I think semantically it is just a container.

Comment (wconstab) [reply train.py]: so what would the proposed flow be?

```
modules = ModuleList(model_parts)

apply_spmd(_) # would we use 'modules' here and deal with applying fsdp one more time to the root, but operating over the first level of children for the current tp/dp logic?  Or would we apply this to model_parts and then manually apply fsdp one more time outside to 'modules'?

checkpoint.save(model_parts)  # i think we can't use the outer wrapper during checkpoint save/load right or have to adjust fqn manually there?
```

Comment (awgu) [reply train.py]: I think on the FSDP side, the main constraint is actually just to not call `fully_shard(ModuleList)` since then FSDP forward hooks would never run (we actually error today with FSDP2 if you try to call `fully_shard(ModuleList)`). If you were to use a `ModuleList`, then we should just apply `fully_shard` to the children of the `ModuleList`.

I am not as familiar with the checkpointing constraints. If using a `ModuleList` causes too much checkpointing headache, then feel free to not use a `ModuleList`. Either way works for FSDP.

File: torchtitan/checkpoint.py [hunk 3]
```diff
@@ -137,13 +137,61 @@ def __init__(
...
         self.states.update(
             {
-                "model": ModelWrapper(model),
-                "optimizer": OptimizerWrapper(model, optimizer),
-                "lr_scheduler": lr_scheduler,
+                "model": ModelWrapper(model_parts),
+                "optimizer": OptimizerWrapper(model_parts, optimizers),
+                # TODO(whc) flatten lr_schedulers using a wrapper and somehow handle resharding?
+                # or store one per key and explicitly dont support resharding?
+                # "lr_scheduler": lr_scheduler,
```
Comment (wconstab) [inline torchtitan/checkpoint.py]: i have to fix this part. i think @fegin had a suggestion for a workaround that would support resharding. i am not sure if i should do this or just let it be saved in a way that would break for resharding and don't care about resharding for now.

File: torchtitan/checkpoint.py [hunk 4]
```diff
@@ -137,13 +137,61 @@ def __init__(
 
         if not self.enable_checkpoint:
             return
+        """
+        Tricky changes for Pipeline Parallelism
+
+        1. even for simple PP schedules, we introduce a problem where there is a separate optimizer on stage1's rank
+        vs stage0's rank.  When saving, these collide and one of them is lost.  Then when reloading, only one stage can
+        restore its optimizer states, others will error.
+            --> no fix yet
```
Comment (wconstab) [inline torchtitan/checkpoint.py]: this mega-comment is out of date. some of these issues are fixed already. ill update the comment with the latest status/issues.

File: torchtitan/checkpoint.py [hunk 5]
```diff
@@ -137,6 +137,38 @@ def __init__(
...
+
+        2. With complex PP schedules, we have multiple model chunks per pp rank. This compounds challenge (1) by also
+        requiring us to reason about multiple 'optim' objects locally.
+
+            We solve this in the Model and Optimizer wrapper classes by flattening the state dicts from each object
+            into one state dict before saving/loading. We rely on the individual state_dicts to not collide,
+            which is gauranteed for the model by correct pipeline splitting and for the optimizer by the flattening
+            support described in (1).
+
+        3. LR schedulers also index model states like optimizers and would need to be flattened properly to support
```
Comment (wconstab) [inline torchtitan/checkpoint.py:160]: RE:
image

I did it this way because i thought if we are not supporting resharding of lr_scheduler, then i may as well save each one.  If i save each one, then at load time i have a form of assertion provided for me by dcp- if the runtime loading the checkpoint has the same number of ranks, they will match up and load OK. if not, they would throw an error.

I could switch back to the version where I save only one copy, but then i have to do some validation up front.  Do you think this way is better?
@fegin

File: torchtitan/parallelisms/parallelize_llama.py [hunk 6]
```diff
@@ -191,67 +197,104 @@ def pipeline_llama_manual(
...
+        rank: int, num_stages: int, style: str = "loop"
+    ) -> Tuple[int]:
+        """Compute the stage ids for the stages that will run on this pp rank for either a looped or V style schedule"""
+        assert (
+            num_stages % pp_size == 0
+        ), f"num_stages {num_stages} must be evenly divisible by pp_size {pp_size}"
+        stages_per_rank = num_stages // pp_size
+        if style == "loop":
+            return tuple(pp_rank + s * pp_size for s in range(stages_per_rank))
+        elif stype == "v":
```
Comment (kwen2501) [inline torchtitan/parallelisms/parallelize_llama.py]: Typo?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 7]
```diff
@@ -191,67 +197,104 @@ def pipeline_llama_manual(
...
+        logger.info(
+            f"PP rank {pp_rank} is building stage_idx {stage_idx}"
+            f" with start_layer {start_layer}, stop_layer {stop_layer}: model chunk \n{model_chunk}"
+        )
+        stages.append(stage)
+        models.append(model_chunk)
+    return stages, models
 
 
 def pipeline_llama_tracer(
```
Comment (kwen2501) [inline torchtitan/parallelisms/parallelize_llama.py:282]: Apply similar changes to the tracer path?

**Review by kwen2501 — APPROVED** [review]
Pipelining part looks good to me. Left two minor comments.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 6 above)
Comment (wconstab) [reply torchtitan/parallelisms/parallelize_llama.py]: yep

File: torchtitan/parallelisms/pipelining_utils.py [hunk 8]
```diff
@@ -26,7 +33,7 @@ def build_pipeline_schedule(job_config, parallel_dims, stages, loss_fn):
         n_microbatches = job_config.experimental.pipeline_parallel_degree
 
     return schedule_class(
-        stage,
+        stages if looped_schedule else stages[0],
```
Comment (donglimm) [inline torchtitan/parallelisms/pipelining_utils.py:38]: the init function of ScheduleInterleaved1F1B takes a list of _PipelineStageBase
if we only pass stages[0], will it cause ScheduleInterleaved1F1B fail ?

Comment (wconstab) [reply torchtitan/parallelisms/pipelining_utils.py:38]: well, the point of this code is that the schedule class could be a simple schedule or a looped schedule.

If it's a looped schedule, like Interleaved1F1B, then we must have 'stages'.

if its a simple schedule, then we must have just one stage, so 'stages[0]' is appropriate.

But we should never have Interleaved1F1B(stages[0]).

File: torchtitan/checkpoint.py (hunk 5 above)
Comment (fegin) [reply torchtitan/checkpoint.py:160]: If we are not supporting resharding, then this implementation is better.
